### ISSUE TITLE
version 1.3 generalize configuration language

### ISSUE THREAD

**Issue by jrchudy** [informatics-isi-edu/deriva-webapps#101]
In Epic issue #95 , we detailed the changes that would be made over time to update the plot configuration language as we were developing the violin plot type. At the end of this, we discussed what the eventual general configuration should be like so different plot types can reuse these definitions. The latest discussion around this "final" configuration language is [here](https://github.com/informatics-isi-edu/deriva-webapps/issues/95#issuecomment-702423938) (ignore the parts that are specific to violin plots, `recordsetSelectors` and `traces[].multiStudy`, which are removed from the example below).

Looking at that configuration, the idea is to align the old configuration language with our changes we made as we were designing violin plots. Plotly takes 3 values as part of the constructor to render data, `data`, `layout`, and `config`. We want to be able to have properties defined in the configuration that are passed directly to plotly and another object of properties that we know the names of to override properties attached to ploty. Previously we had 2 objects named `config` and `plotly`:
  - We want to change `config` to be `plotly.config`. We want to keep the name of the property consistent with what plotly expects. Since in the code the plotly object is `plotly: { data: [{...}, ...], config: {...}, layout: {...} }`, we want our configuration to be the same. This also allows for layout (and even data) to be added under `plotly`
  - We also want to change `plotly_config` to just be `config`. Everything under config will be iterated over and used to override properties in either `plotly.config` or `plotly.layout`.

```
var plotConfigs = {
    "config-name": {
        plots: [
            {
                plot_type: type,
                // "apply as is"
                plotly: {
                    // passed directly to plotly (plotly.config)
                    config: {
                        modeBarButtonsToRemove: ["select2d", "lasso2d", "hoverClosestCartesian", "hoverCompareCartesian", "toggleSpikelines"],
                        displaylogo: false,
                        responsive: true
                    }, 
                    layout: {...},
                    data: [ {...}, ... ]
                },
                // configuration options used to override plotly definitions
                config: {
                    title_display_pattern: "",
                    ...
                    xaxis: {
                        title_display_pattern: "",
                        tick_display_pattern: "",
                        ...
                    },
                    yaxis: {
                        title_display_pattern: "",
                        tick_display_pattern: "",
                        ...
                    },
                },
                ...
            }
        ]
    }
}
```

Changes to make:
 - [x] update the code to honor these configuration name changes (`config => plotly.config` and `plotly_config => config`)
   - needs to be done for each plot type (check switch statements in code)
   - should already be done for violin plot code path
 - [x] update each configuration in plot-config.js
   - I suggest copying each configuration and renaming the copied one. Something like:
      ```
      "gudmap-todate-bar": { ... },
      "gudmap-todate-bar-v3": { ... }
      ```
   - change object names as described above (`config => plotly.config` and `plotly_config => config`)
   - rearrange properties according to whether they are setting or overriding a plotly value
     - a lot of the values in the old `plotly_config` are applied directly to the `layout` property, so most of these can be moved to the new `plotly.layout` instead of staying in the new `config` object.
 - [x] test each configuration change against the same page on master branch to ensure nothing was degraded
 - [x] propose configuration changes for rbk `plot-config.js` to align with the potential code changes

An issue with violin plots that was added here before:
 - [x] make the tooltip disappear if the button is clicked, so it won't block the drop-down text on group by and scale

**Comment by DevanshiDesai19** [comment]
The following changes are made to the configuration file:

"gudmap-todate-pie-v3": {
    page_title: "GUDMAP Release Status Dashboard",
    // Array of object plots to be shown on the page
    plots: [
        {
            // ValuesValues can be from : "line", "bar", "dot", "area", "dot-lines", "pie", "histogram-horizontal", "histogram-verical"
            plot_type: "pie",
              plotly:{
                  layout: {
                    **title: "Number of GUDMAP resources released to-date", //  plot_title is moved here**
                    **showlegend: true,  // Moved  from config object**        
                  }
              },
              config:{
                **title_hyperlink_pattern: "[{{Title}}](https://dev.isrd.isi.edu/chaise/search{{Context_Params}}){target=_blank}",  //Added as a part of Update plot app to allow for text links#106 issue**
                **slice_label: "value",     // Moved from plotly.config to here.**
                **format_data: true,         // Moved from plotly.config to here.**
                **disable_default_legend_click: true //Added as a part of Update plot app to allow for text links#106 issue**
              },
            traces: [
                {
                    // The request url that has to be used to fetch the data.
                    uri: "/ermrest/catalog/2/entity/M:=Dashboard:Release_Status/Consortium=GUDMAP/!(%23_Released=0)/!(Data_Type=Antibody)/!(Data_Type::regexp::Study%7CExperiment%7CFile)/$M@sort(ID::desc::)?limit=26", //Added as a part of Update plot app to allow for text links#106 issue**
                    // legend: ["Browser All Events 1","Browser All Events 2"],   // name of traces in legend
                    data_col: "#_Released",     // name of the attribute of the data column
                    legend_col: "Data_Type",     // name of the attribute of the legend column
                    show_percentage: false,     // to show the percentage or not on slices
                    //lengend_template_params:'{"Schema_Table": "row[\'Schema_Table\']","Data_Type_Filter": "ERMrest.encodeFacet(row[\'Data_Type_Filter\'])","Title": "traceLabel"}',
                    legend_markdown_pattern:"[{{Devanshi}}](/chaise/recordset/#2/{{Schema_Table}}/*::facets::{{Data_Type_Filter}}){target=_blank}",  //Added as a part of Update plot app to allow for text links#106 issue**
                }
            ],
            // Plotly defualt buttons/actions to be removed
            plotlyDefaultButtonsToRemove: ["scrollZoom", "zoom2d","sendDataToCloud","autoScale2d", "lasso2d", "select2d", "hoverClosestCartesian", "hoverCompareCartesian", "toggleSpikelines", "hoverClosestPie"]
        }
    ]
},

**Comment by DevanshiDesai19** [comment]
"gudmap-todate-bar-v3": {
    page_title: "GUDMAP Data Status Dashboard",
    // Array of object plots to be shown on the page
    plots: [
        {
            plot_type: "bar",
            plotly:{
                layout: {
                    **title: "Number of GUDMAP resources released to date (log scale)", //plot_title moved here**
                    **height: 500, // moved here from config object
                    width: 1200, // moved here from config object
                    xaxis: {
                        title: "Number of Records",          // moved here from config object
                        type: 'log',             // optional value: tickformat should compatible with type // moved here from config object
                    },
                    yaxis: {
                        // title: "Data_Types",       // plot y_axis label // moved here from config object
                    },
                    margin: {
                        t: 30,  // moved here from config object
                        l: 280  // moved here from config object
                    },
                    legend:{
                        traceorder: "reversed"    // order of the legend is reversed // moved here from config object
                    },**   
                }
            },
            config:{
                **title_hyperlink_pattern: "[{{Title}}](https://dev.isrd.isi.edu/chaise/search{{Context_Params}}){target=_blank}", //Added as a part of Update plot app to allow for text links#106 issue**
                **format_data_x: true,   // Moved from plotly.config**
                xaxis: {
                    **title_hyperlink_pattern: "[{{Title}}](https://dev.isrd.isi.edu/chaise/search{{Context_Params}}){target=_blank}", //Added as a part of Update plot app to allow for text links#106 issue**
                    // tickformat: ',d',     // format for the ticks. For more formatting types, see: https://github.com/d3/d3-format/blob/master/README.md#locale_format
                },
            },
            traces: [
                {
                    // The request url that has to be used to fetch the data.
                    uri: "/ermrest/catalog/2/entity/M:=Dashboard:Release_Status/Consortium=GUDMAP/!(%23_Released=0)/!(Data_Type=Antibody)/!(Data_Type::regexp::Study%7CExperiment%7CFile)/$M@sort(ID::desc::)?limit=26",
                    legend: ["#_Released"],   // name of traces in legend
                    x_col: ["#_Released"],    // column name to use for x values
                    y_col: ["Data_Type"],                   // array of column names to use for y values
                    orientation: "h",                      // Optional parameter for displaying the bar chart horizontally
                    **y_col_hyperlink_pattern: "/chaise/recordset/#2/{{Schema_Table}}/*::facets::{{Data_Type_Filter}}", //Added as a part of Update plot app to allow for text links#106 issue**
                    // hovertemplate: ',d',
                    textfont: {
                        size: 10                           // It will work till the bar size can accomodate the font size
                    },
                },
            ],
            // Plotly defualt buttons/actions to be removed
            plotlyDefaultButtonsToRemove: ["scrollZoom", "zoom2d","sendDataToCloud","autoScale2d", "lasso2d", "select2d", "hoverClosestCartesian", "hoverCompareCartesian", "toggleSpikelines"]
        }
    ]
},

**Comment by jrchudy** [comment]
This issue is resolved with PR #117 being merged.
